### DISCUSSION TITLE
Welcome to pygbag Discussions!

### DISCUSSION THREAD

**Discussion by pmp-p** [pygame-web/pygbag#1] [Announcements]
Local rules :
 - Get the things done !
 - Follow KISS principles.
 - Keep it **FUN**, although it packs cpython it's for making games !

**Comment by tom-denesyk** [comment]
Pygbag newbie here.
Is there something wrong with the way a pygbagged app works with sockets? Or maybe with my simple client\server example?

Here's a main() that establishes socket connection to a server at localhost, port=2222.
main() echoes to server msgs received from server.
This simple client/server works when main() is executing in two dos windows as python programs.
Does not work when main() is running under pygbag as a web app, ie 'pygbagged'. 

[main.zip](https://github.com/pygame-web/pygbag/files/10203053/main.zip)

**Reply by pmp-p** [reply]
you need a websocket proxy in front of the echo service for the browser to connect to. Browsers are not allowed to connect plain desktop sockets.
here's such a proxy https://github.com/novnc/websockify

**Reply by tom-denesyk** [reply]
Hi Paul,
Thanks for your earlier response.

I now have my code using websockets . 
Without pygbag communications between client and server are working as expected.

However, I’m using a python virtual environment and I think this is causing a problem…

When I open a webrowser localhost:8000#debug I get this… (See line with <<<<< *****)

****************************************
['noodling', 'sys', 'asyncio', 'websockets', 'pygame']
****************************************
Traceback (most recent call last):
  File "/data/data/org.python/assets/site-packages/aio/__init__.py", line 239, in step
    fn(*argv, **kw)
  File "<string>", line 193, in execfile
  File "/data/data/noodlingwebsocket/assets/main.py", line 1, in <module>
    from noodling import *
  File "/data/data/noodlingwebsocket/assets/noodling.py", line 14, in <module>
    import websockets
ModuleNotFoundError: No module named 'websockets'    <<<<< *****    

Any idea?

Thanks,
Tom

**Reply by pmp-p** [reply]
you don't need to use websockets in python, the emscripten "os" layer is already converting browser websocket to normal socket internally for python. here's an example connecting to irc ( with websockify  as proxy ) https://github.com/pygame-web/showroom/blob/main/pygame-scripts/aio.sockets.html

**Comment by tom-denesyk** [comment]
Sorry, I don't understand what to do with 'example connecting to irc...'.
In the client/server example what would replace the use of sockets?
Can you point me at an example where a pygbagged client communicates with a python server?
Regards,
Tom

**Reply by tom-denesyk** [reply]
Ah. 
Maybe this is just such an example?
https://docs.python.org/3/library/asyncio-stream.html#asyncio-streams

I'll give this a try.
Cheers,
Tom

**Reply by pmp-p** [reply]
@tom-denesyk if you really don't know what to do i suggest you join pygame discord.

 Also i won't provide any support for asyncio socket/streams from stdlib because asyncio is not suitable for game loops, 

uasyncio maybe  but only if i have time for that and if pyscript does not take on that before (and they should).


The aio.sockets.html page ( read its python source ) is exactly connecting a python wasm client to a python server ( in that case it is https://github.com/pmp-p/uircd  + websockify )

**Comment by Digital-Physics** [comment]
I was very happy to find pygbag, although I think I might need to do a lot of refactoring of my PyGame code. Getting rid of yaml, numpy, etc. imports is doable. But I have a main game loop with functions that are occasionally called, which run in their own nested game loops. These self-contained loops, which correspond to different rooms, run several steps until the loop is exited, at which point they pop back up to the main game loop level. Is there a way to insert async-and-await code across my game files which contain these nested game loop functions and still get pygbag to work? Or will I need to extract the logic in these nested game loop functions and put it all on the top game loop level (no nested loops) to get pygbag to work? Thanks!

**Reply by pmp-p** [reply]
nested loop are bad, use only one loop and game states : see an example of code being refactored from nested loop to state change ( not finished ) here https://github.com/pmp-p/pygame-escape_the_lab-wasm/blob/main/main.py

**Reply by Digital-Physics** [reply]
I had a feeling that was the case. :( Thanks for the quick reply and the reference! I'm gonna make it work!

**Reply by pmp-p** [reply]
btw numpy is ok for 3.11

**Reply by Digital-Physics** [reply]
cool, I will try it in 3.11 and see if importing numpy doesn't error out. That would save me refactoring a good bit of code.

**Reply by pmp-p** [reply]
for numpy testing use  `--template ctx3d.tmpl --cdn https://pygame-web.github.io/showroom/dev/` and put `import numpy`  and `import pygame` in the main.py.  that version is making dynamic import from reading main.py  ( it will be pygbag 0.6 soon )

**Reply by Michael-Everest** [reply]
edit - not completly sure on this but...

im pretty sure i know your problem and i have had the same one. i think this code is what your looking for?
short answer, just put await in front of the function call and it works like normal
- ex await output = this_is_a_function(var1, var2)
- and make sure to put async infront of def where the function is defined
- async def this_is_a_function(var1, var2):

**Comment by Digital-Physics** [comment]
Second question, while I'm here :) ... memory management... Are there optimization techniques I should use to avoid the memory error I get when I load a lot of pngs at the start of my game? Perhaps if I load this data right before I need it, and then clear it from memory afterwards using something like "with open()", I will be ok?

**Reply by pmp-p** [reply]
To avoid eating memory use a 1024x600 screen resolution and resize your sprites accordingly. If you have too much sprites i don't know maybe use a cache but i'm not sure  how and when sprites memory will be released ( if it is ). That would need testing .

**Reply by Digital-Physics** [reply]
ok, I'll experiment. Thanks again!

**Reply by Digital-Physics** [reply]
Update: I was able to refactor my PyGame code and incorporate pygbag. Thanks again for creating this pivotal library! "Digital Physics" the video game will be released on itch.io on Groundhog Day, February 2, and thanks to pygbag, people will be able to play it in the browser! I'll post the link when it is released. Thanks again!

**Reply by Digital-Physics** [reply]
Ok, here's my "Digital Physics" game built with pygbag 0.5, python 3.10.8, pygame 2.1.2. I'm really happy with the deployment results, so thanks again! I look forward to seeing how future versions of the pygbag library evolve. 

https://digital-physics.itch.io/digital-physics

Feel free to try the game out, offer feedback, and share it with people. Thanks again!

**Comment by pmp-p** [comment]
To allow multiple and more clear threads, please use https://github.com/pygame-web/pygbag/discussions/new/choose to create new subjects from now on
